### ISSUE TITLE
New option: timestampleft

### ISSUE THREAD

**Issue by tyralla** [hydpy-dev/hydpy#84]
There is often confusion with the position of interval timestamps.  In HydPy, we mostly solve this problem by defining [Timegrid](https://hydpy-dev.github.io/hydpy/4.0/timetools.html#hydpy.core.timetools.Timegrid) objects.  We place, for example, a text defining a [Timegrid](https://hydpy-dev.github.io/hydpy/4.0/timetools.html#hydpy.core.timetools.Timegrid) object into the header of each ASCII time-series file.  In `Timegrid("2000-01-01", "2000-01-03", "1d")`, `"2000-01-01"` is the start date and `"2000-01-03"` is the end date.  Then, for single precipitation time series, there must be two file entries, let's say "x1" for the first interval (from 2000-01-01 00:00 to 2000-01-02 00:00) and "x2" for the second interval (from 2000-01-02 00:00 to 2000-01-03 00:00).  So, for interval related properties like precipitation, everything is clear.  However, some related issues need to be improved or at least clarified.

The interval-related ASCII file format does not fit well with properties calculated for specific time points (most importantly, the values of [StateSequence](https://hydpy-dev.github.io/hydpy/4.0/sequencetools.html#hydpy.core.sequencetools.StateSequence) instances, sometimes also the values of [FactorSequence](https://hydpy-dev.github.io/hydpy/4.0/sequencetools.html#hydpy.core.sequencetools.FactorSequence) like [WaterLevel](https://hydpy-dev.github.io/hydpy/master/dam.html#hydpy.models.dam.dam_factors.WaterLevel)).

Iteration over a [Timegrid](https://hydpy-dev.github.io/hydpy/4.0/timetools.html#hydpy.core.timetools.Timegrid) object yields the left timestamps of all intervals:

```python
>>> for date, x in zip(Timegrid("2000-01-01", "2000-01-03", "1d"), ("x1", "x2")): print(date, x)
2000-01-01 00:00:00 x1
2000-01-02 00:00:00 x2
```

When working with NetCDF files, we assume (when reading data) or define (when writing data) that each timestamp corresponds to the beginning of the respective interval:

```python
>>> Timegrid("2000-01-01", "2000-01-03", "1d").firstdate.to_cfunits()
'hours since 2000-01-01 00:00:00 +01:00'
>>> Timegrid("2000-01-01", "2000-01-03", "1d").to_timepoints()
array([ 0., 24.])
```

Our idea is to allow more flexibility by introducing the new boolean option `timestampleft` (or `timestamp_left`).  This option should at least modify reading and writing NetCDF files (to improve compatibility with Delft-FEWS).  

Open questions:

- Should the new option also affect the related features of the [Timegrid](https://hydpy-dev.github.io/hydpy/4.0/timetools.html#hydpy.core.timetools.Timegrid) class (like iteration) in a more general manner?
- Should we somehow modify the ASCII format when writing state series?
- Are there other *HydPy* features with similar, timestamp-related unclarities?

**Comment by tyralla** [comment]
Commit 6143423b82fe648ec65ea993d0ede4490186213f add the `timestampleft` option and commit b8da4b6d06b5aa50be2b609957ef92c6d9ff57ab uses it to achieve the desired flexibility in reading and writing NetCDF files.
  
Additionally, *HydPy* now writes additional metadata into each NetCDF file, either `timestampleft = True` or `timestampleft = False`.   And it uses this information when reading these files.  If the information is missing, it falls back to the current value of the introduced "global" option, which one can, for example, define in the usual way via XML configuration files.  If the file-specific information is missing or in conflict with the global option, it emits a warning.  (Hopefully, other tools like Delft-FEWS allow to add such metadata.)

> Should the new option also affect the related features of the [Timegrid](https://hydpy-dev.github.io/hydpy/4.0/timetools.html#hydpy.core.timetools.Timegrid) class (like iteration) in a more general manner?

Implemented by commit b60159851154f7317eb59423ad715ade922055c6.  However, I am not completely sure if this is a real improvement.  Let us try it for a while and revert it if it reveals to be more confusing than helpful.

**Comment by tyralla** [comment]
> Should we somehow modify the ASCII format when writing state series?

Still open for discussion.  

Two potential problems: 
- `Timegrid` objects with zero lengths are currently not allowed.
- Some `FactorSequence` classes need handling like `FluxSequence` classes and others like `StateSequence` classes (are there other exceptions?).

**Comment by tyralla** [comment]
> If the file-specific information is missing or in conflict with the global option, it emits a warning.  (Hopefully, other tools like Delft-FEWS allow to add such metadata.)

Delft-FEWS does not seem to support this.  Hence, we removed the warning for missing file-specific information in commit 03a78572b4c719a39349ae61951c9be0a165fae7.

> Should we somehow modify the ASCII format when writing state series?

We decided against it, as the current handling is not wrong, and there is no obvious way to make it more explicit while keeping its correctness in all cases.

However, there is a related problem concerning NetCDF files.   Commit 03857fa44497685fa5fed3446c44801a848067a3 replaces the NetCDF file attribute `timestampleft` with `timereference` with the allowed values `left interval boundary`, `right interval boundary`, and `current time`.   The first two are relevant for "time interval sequences" (e.g. precipitation) and the last one for "time point sequences" (e.g. soil moisture).  The HydPy option (or XML option) `timestampleft` remains unchanged.  See the commit message and the documentation on the [query_timegrid](https://hydpy-dev.github.io/hydpy/master/netcdftools.html#hydpy.core.netcdftools.query_timegrid) function for more detailed information (Travis cannot update the online documentation at the moment - there seems to be an issue with installing *sphinxcontrib* - I will have a look at it later).

I think if no one finds any pitfalls in our new handling of timestamps in NetCDF files, we can close this issue.
